pkg/protocol/commitment.go

```diff
@@ -0,0 +1,285 @@
+	}
+
+	return nil
+}
+
+// initLogger initializes the Logger of this Commitment.
+func (c *Commitment) initLogger() (shutdown func()) {
+	c.Logger, shutdown = c.commitments.NewEntityLogger(fmt.Sprintf("Slot%d.", c.Slot()))
+
+	return lo.Batch(
+		c.Parent.LogUpdates(c, log.LevelTrace, "Parent", (*Commitment).LogName),
+		c.MainChild.LogUpdates(c, log.LevelTrace, "MainChild", (*Commitment).LogName),
+		c.Chain.LogUpdates(c, log.LevelTrace, "Chain", (*Chain).LogName),
+		c.RequestAttestations.LogUpdates(c, log.LevelTrace, "RequestAttestations"),
+		c.WarpSyncBlocks.LogUpdates(c, log.LevelTrace, "WarpSyncBlocks"),
+		c.Weight.LogUpdates(c, log.LevelTrace, "Weight"),
+		c.AttestedWeight.LogUpdates(c, log.LevelTrace, "AttestedWeight"),
+		c.CumulativeAttestedWeight.LogUpdates(c, log.LevelTrace, "CumulativeAttestedWeight"),
+		c.IsRoot.LogUpdates(c, log.LevelTrace, "IsRoot"),
+		c.IsAttested.LogUpdates(c, log.LevelTrace, "IsAttested"),
+		c.IsVerified.LogUpdates(c, log.LevelTrace, "IsVerified"),
+		c.ReplayDroppedBlocks.LogUpdates(c, log.LevelTrace, "ReplayDroppedBlocks"),
+		c.IsEvicted.LogUpdates(c, log.LevelTrace, "IsEvicted"),
+
+		shutdown,
+	)
+}
+
+// initDerivedProperties initializes the behavior of this Commitment by setting up the relations between its properties.
+func (c *Commitment) initDerivedProperties() (shutdown func()) {
+	return lo.Batch(
+		// mark commitments that are marked as root as verified
+		c.IsVerified.InheritFrom(c.IsRoot),
+
+		// mark commitments that are marked as verified as attested
+		c.IsAttested.InheritFrom(c.IsVerified),
+
+		c.Parent.WithNonEmptyValue(func(parent *Commitment) func() {
+			// the weight can be fixed as a one time operation (as it only relies on static information from the parent
+			// commitment)
+			c.Weight.Set(c.CumulativeWeight() - parent.CumulativeWeight())
+
+			return lo.Batch(
+				parent.deriveChildren(c),
+
+				c.deriveChain(parent),
+				c.deriveCumulativeAttestedWeight(parent),
+				c.deriveIsAboveLatestVerifiedCommitment(parent),
+
+				c.Chain.WithNonEmptyValue(func(chain *Chain) func() {
+					return lo.Batch(
+						c.deriveRequestAttestations(chain, parent),
+						c.deriveWarpSyncBlocks(chain, parent),
+					)
+				}),
+			)
+		}),
+
+		c.Chain.WithNonEmptyValue(func(chain *Chain) func() {
+			return lo.Batch(
+				chain.addCommitment(c),
+
+				c.deriveReplayDroppedBlocks(chain),
+			)
+		}),
+	)
+}
+
+// deriveChildren derives the children of this Commitment by adding the given child to the Children set.
+func (c *Commitment) deriveChildren(child *Commitment) (unregisterChild func()) {
+	c.MainChild.Compute(func(mainChild *Commitment) *Commitment {
+		if !c.Children.Add(child) || mainChild != nil {
+			return mainChild
+		}
+
+		return child
+	})
+
+	return func() {
+		c.MainChild.Compute(func(mainChild *Commitment) *Commitment {
+			if !c.Children.Delete(child) || child != mainChild {
+				return mainChild
+			}
+
+			return lo.Return1(c.Children.Any())
```

Comment:
Why can we just take any child here? Shouldn't it be the heaviest?

Reply:
It doesn't really matter which chain is labeled how as the chain switching logic will eventually correct any mislabeling, but you are right that we could pre-select the heaviest child to have less state changes in case we happen to pick the wrong one first.

These kind of reorgs are so seldom that it felt less complicated in code to just leave the selection mechanism to the chainswitching part and just make sure things are uniquely labeled here.